### PR TITLE
extract stash explorer

### PR DESCRIPTION
**Pull request by rebornix**
Move stash explorer out of Gitlens explorer.

### CODE REVIEW

File: src/views/gitStashExplorer.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+import { Event, EventEmitter, ExtensionContext, TreeDataProvider, TreeItem, Uri, window } from 'vscode';
+import { ExplorerNode, StashNode } from './gitExplorerNodes';
+import { GitService, GitUri } from '../gitService';
+import { StashCommitNode } from './stashCommitNode';
+
+export * from './gitExplorerNodes';
+
+export class GitStashExplorer implements TreeDataProvider<ExplorerNode>  {
```
Comment (eamodio) [inline src/views/gitStashExplorer.ts]: Let's just call it StashExplorer -- I know my naming has git everywhere, but I'm going to try to move away from it eventually ;)

File: src/views/gitStashExplorer.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    constructor(private context: ExtensionContext, private git: GitService) {
+        const editor = window.activeTextEditor;
+
+        const uri = (editor !== undefined && editor.document !== undefined)
+            ? new GitUri(editor.document.uri, { repoPath: git.repoPath, fileName: editor.document.uri.fsPath })
+            : new GitUri(Uri.file(git.repoPath), { repoPath: git.repoPath, fileName: git.repoPath });
+
+        this._node = new StashNode(uri, this.context, this.git);
```
Comment (eamodio) [inline src/views/gitStashExplorer.ts]: What do you think of keeping this in a node or just nuking the node and dropping the logic in the explorer? I keep going back and forth -- eh keeping it in the node is nice for possible reuse later in another view or whatnot

File: src/views/gitStashExplorer.ts (hunk 1 above)
Comment (rebornix) [inline src/views/gitStashExplorer.ts]: Addressed.

File: src/views/gitStashExplorer.ts (hunk 2 above)
Comment (rebornix) [inline src/views/gitStashExplorer.ts]: The first time I saw the code, I found that it's not intuitive enough. For example, let's say we have this Stash Explorer, the tree node can only be `StashNode` or `FileChangeNode` so a union type `StashNode | FileChangeNode` is good enough. That's also how I implemented it.

But for your case, if the hierarchy can be more than 2, your solution is definitely smarter. The leaf node is always `FileChangeNode` in Commits, Stashes, Pull Request so I'll say let's go with this way, putting things into the node and reuse them with ease.

**Pull request merged**